### ISSUE TITLE
Move back out of tools.wmflabs.org

### ISSUE THREAD

**Issue by hartman** [Pathoschild/Wikimedia-contrib#113]
Due to the enablement of CSP, and the stricter code loading policies that in the future will start being implemented, I suggest moving this code back into meta.wikimedia.org, as tools.wmflabs.org is not a trusted domain within the WM ecosystem.
See also https://en.wikipedia.org/wiki/Wikipedia:Village_pump_(technical)#Huge_Script_errors_in_console_log

**Comment by Pathoschild** [comment]
That's a pretty significant change. While it's trivial to copy the scripts to Meta (presumably in the MediaWiki namespace), there are countless references to the scripts across all wikis that would need to be updated.

**Comment by hartman** [comment]
Yeah, let's also be a bit careful here in terms of effort. There are some ideas of alternative solution strategies. Might be best to privately communicate with Bawolff .

**Comment by MarcoAurelioWM** [comment]
I agree this is a very important decision to make. I feel this script suite is better maintained here, but maybe to bypass CSP we could get Phabricator Diffusion (phabricator.wikimedia.org) to keep a clone of this repo and have `mw.loader.load`ed from there? If we want to explore that path, I can create repositories on Phabricator and try to test that.

**Comment by Pathoschild** [comment]
@MarcoAurelioWM  That might be a good middle approach. That lets us keep deploying updates through Git, but WMF has the ability to freeze or remove the mirror if needed. Would Diffusion mirror the upstream repo automatically, or would someone need to do it manually each time?

**Comment by MarcoAurelioWM** [comment]
@Pathoschild Indeed you can keep the tool development here in GitHub and just ask Phabricator Diffusion to watch this GitHub repo for updates. In that case Phabricator Diffusion will simply watch from time to time, automatically, if there have been any updates upstream (this GitHub repository) and if there's any, Phabricator will import them.

For example, I keep the configuration files of my bot here at GitHub (https://github.com/MarcoAurelioWM/MABot-jobs) and I have a mirror of said code at Phabricator (https://phabricator.wikimedia.org/diffusion/TMAB/). Once I merge some code at my GitHub repo, Phabricator will automatically check if there has been updates and will import the code.

Like I've said, the updates happen automatically but _not_ immediatelly. Phabricator does fetch updates from repos depending on its activity, to better spend its resources I think. That is, for the most active repositories (for example, MediaWiki core) Phabricator tries to fetch updates each 30 seconds. For more quiet repositories, Phabricator can take up to 6 hours to go and look if there has been any change. You can, however, order Phabricator to update the repository at any given time using a button in the repository.

Phabricator can also be used to host repositories locally as we use Gerrit or GitHub. In that case updates happen directly on push, but Phabricator is --IMHO-- currently not very user friendly when it comes to use it as a platform for code review and repository hosting.

Hope that this helps. In case of further questions, please let me know. Thanks.

**Comment by MarcoAurelioWM** [comment]
One problem so far, however, is that I'm not sure if Phabricator "raw" files can be used. Last time I tried to use a "raw" file the link expired after some time, rendering the link useless. If that is still happening then there's no point in trying to mirror this code there. I'll ask around.

**Comment by MarcoAurelioWM** [comment]
As far as I could see the problem still exists. File links are all temporary thus cannot be used as a solution for this I'm afraid 😿

**Comment by MarcoAurelioWM** [comment]
Yes. I agree, but it seems that the static resources served via
phab.wmfusercontent.org do all expire after some time. Plus it is also very
possible that wmfusercontent is also CSP-blocked so we wouldn't achieve the
CSP bypassing.

>

**Comment by Pathoschild** [comment]
@MarcoAurelioWM Have you heard anything back from the tech team on the above question? (Also, wouldn't it make sense to just allow `*.wmflabs.com` in the policy, since it's a Wikimedia domain too?)

**Comment by MarcoAurelioWM** [comment]
Hello. I've not heard anything, nope. I can try to ask again if that helps.

**Comment by Pathoschild** [comment]
This was brought up again in Phabricator tickets [T304151](https://phabricator.wikimedia.org/T304151) and [T304208](https://phabricator.wikimedia.org/T304208). Seems moving the scripts to wiki pages is still the only solution, so I started [a discussion on Meta](https://meta.wikimedia.org/wiki/Tech#Preferred_location_for_crosswiki_site_scripts%3F) about the best place to put them.

**Comment by 20after4** [comment]
The wmfusercontent domain is intentionally separate from phabricator.wikimedia.org and allowing it in the csp would defeat the purpose of having a separate domain.  The reason for it to exist is to avoid XSS exposure due to allowing users uploading malicious code to Phabricator.
